harvester/DBInterface.py

```diff
@@ -77,107 +69,199 @@ def _prep(self, statement):
-				cur.execute(self._prep("DELETE from subjects where local_identifier = ? and repository_url = ?"), (record['local_identifier'], record['repository_url']))
-				cur.execute(self._prep("DELETE from rights where local_identifier = ? and repository_url = ?"), (record['local_identifier'], record['repository_url']))
-				cur.execute(self._prep("DELETE from descriptions where local_identifier = ? and repository_url = ?"), (record['local_identifier'], record['repository_url']))
+				cur.execute(self._prep("DELETE from creators where record_id=?"), (record['record_id'],))
+				cur.execute(self._prep("DELETE from subjects where record_id=?"), (record['record_id'],))
+				cur.execute(self._prep("DELETE from rights where record_id=?"), (record['record_id'],))
+				cur.execute(self._prep("DELETE from descriptions where record_id=?"), (record['record_id'],))
+				cur.execute(self._prep("DELETE from tags where record_id=?"), (record['record_id'],))
+				cur.execute(self._prep("DELETE from geospatial where record_id=?"), (record['record_id'],))
+				cur.execute(self._prep("DELETE from domain_metadata where record_id=?"), (record['record_id'],))
 			except:
-				self.logger.error("Unable to delete related table rows for record %s in repository %s" % (record['local_identifier'], record['repository_url'] ) )
+				self.logger.error("Unable to delete related table rows for record %s" % (record['local_identifier'],) )
 				return False
 
-		self.logger.debug("Marked as deleted: record %s in repository %s" % (record['local_identifier'], record['repository_url'] ) )
+		self.logger.debug("Marked as deleted: record %s" % (record['local_identifier'],) )
 		return True
 
+	def get_record_id(self, repo_id, local_identifier):
+		returnvalue = 0
+		con = self.getConnection()
+		with con:
+			if self.dbtype == "sqlite":
+				con.row_factory = self.getRow()
+			cur = con.cursor()
+			if self.dbtype == "postgres":
+				from psycopg2.extras import RealDictCursor
+				cur = con.cursor(cursor_factory = RealDictCursor)
+			cur.execute(self._prep("select record_id from records where local_identifier=? and repository_id = ?"),(local_identifier, repo_id))
+			if cur is not None:
+				records = cur.fetchall()
+			else:
+				return 0
+			for record in records:
+				returnvalue = int(record['record_id'])
+		return returnvalue		
 
-	def write_record(self, record, repository_url, metadata_prefix, domain_metadata, mode = "insert"):
+	def write_record(self, record, repo_id, metadata_prefix, domain_metadata, mode = "insert"):
 		if record == None:
 			return None
+		record["record_id"] = self.get_record_id(repo_id, record["identifier"])
 
 		con = self.getConnection()
 		with con:
 			cur = con.cursor()
-			verb = "INSERT"
-			if mode == "replace":
-				verb = "REPLACE"
+			source_url = ""
+			if 'dc:source' in record:
+				source_url = record["dc:source"]
+			if record["record_id"] == 0:
```

Comment:
I'm a little ambivalent about returning 0's for repo/repository ID lookups -- None's should be faster and more logical to test for?